accord-core/src/main/java/accord/primitives/Deps.java

```diff
@@ -0,0 +1,1095 @@
+            // in tx order, in order to map from sorted -> insert order, need "txnIdMap" as the bridge
+            int[] txnIdMap = new int[txnIds.length];
+            for (int i = 0 ; i < txnIdMap.length ; i++)
+                txnIdMap[txnIdLookup.get(txnIds[i])] = i;
+
+            int keyCount = 0;
+            // Key -> distinct([*TxnID]])
+            // at the key position is the OFFSET into the array where the txnIds are for the key
+            // Example
+            // OFFSET(3), OFFSET(5), *tx1, *tx1, *tx2
+            int[] result; {
+                int count = 0;
+                for (int i = 0 ; i < keys.size() ; ++i)
+                {
+                    keyCount += keysToTxnIdCounts[i] > 0 ? 1 : 0;
+                    count += keysToTxnIdCounts[i];
+                }
+                result = new int[count + keyCount];
+            }
+
+            int keyIndex = 0;
+            int offset = keyCount;
+            for (int i = 0 ; i < keys.size() ; ++i)
+            {
+                if (keysToTxnIdCounts[i] > 0)
+                {
+                    int count = keysToTxnIdCounts[i];
+                    int[] src = keysToTxnId[i];
+                    // store all txnIds to results at offset
+                    for (int j = 0 ; j < count ; ++j)
+                        result[j + offset] = txnIdMap[src[j]];
+                    Arrays.sort(result, offset, count + offset);
+                    // the same txnId may be present multiple times, attempt to detect this and push all dups to the end
+                    int dups = 0;
+                    for (int j = offset + 1 ; j < offset + count ; ++j)
+                    {
+                        if (result[j] == result[j - 1]) ++dups;
+                        // positions i == i - 1 but dups were found; replace the "first" dup with the current value, removing it
+                        else if (dups > 0) result[j - dups] = result[j];
+                    }
+                    result[keyIndex] = offset += count - dups;
+                    ++keyIndex;
+                }
+            }
+            if (offset < result.length)
+                result = Arrays.copyOf(result, offset);
+
+            Keys keys = this.keys;
+            if (keyCount < keys.size())
+            {
+                keyIndex = 0;
+                Key[] newKeys = new Key[keyCount];
+                for (int i = 0 ; i < keys.size() ; ++i)
+                {
+                    if (keysToTxnIdCounts[i] > 0)
+                        newKeys[keyIndex++] = keys.get(i);
+                }
+                keys = Keys.of(newKeys);
+            }
+
+            return new Deps(keys, txnIds, result);
+        }
+    }
+
+    public static Builder builder(Keys keys)
+    {
+        return new Builder(keys);
+    }
+
+    private static class LinearMerger extends ArrayBuffers.SavingManager implements DepsConstructor<Object>
```

Comment:
I added a patch to let me disable the cache logic and ran benchmarks with and without buffers... what I see is right now we don't get a win... so I feel we should drop the whole ArrayBuffers class, the logic in LinearMerge seems enough to make sure we can reuse between `Deps`

```
Benchmark[name=merge w/cache(size=15, keys=296, total=6994)] avg 487.11ms, allocated 6028 MiB
Benchmark[name=merge wo/cache(size=15, keys=296, total=6994)] avg 482.43ms, allocated 6028 MiB

Benchmark[name=merge w/cache(size=13, keys=355, total=5447)] avg 365.89ms, allocated 4530 MiB
Benchmark[name=merge wo/cache(size=13, keys=355, total=5447)] avg 364ms, allocated 4530 MiB

Benchmark[name=merge w/cache(size=5, keys=184, total=2435)] avg 45.33ms, allocated 1384 MiB
Benchmark[name=merge wo/cache(size=5, keys=184, total=2435)] avg 45.23ms, allocated 1384 MiB

Benchmark[name=merge w/cache(size=14, keys=335, total=5918)] avg 317.15ms, allocated 4006 MiB
Benchmark[name=merge wo/cache(size=14, keys=335, total=5918)] avg 323.47ms, allocated 4006 MiB

Benchmark[name=merge w/cache(size=18, keys=362, total=9583)] avg 679.28ms, allocated 12133 MiB
Benchmark[name=merge wo/cache(size=18, keys=362, total=9583)] avg 668.39ms, allocated 12692 MiB

Benchmark[name=merge w/cache(size=5, keys=237, total=3162)] avg 91.93ms, allocated 1731 MiB
Benchmark[name=merge wo/cache(size=5, keys=237, total=3162)] avg 83.33ms, allocated 1731 MiB

Benchmark[name=merge w/cache(size=3, keys=149, total=1662)] avg 20.64ms, allocated 807 MiB
Benchmark[name=merge wo/cache(size=3, keys=149, total=1662)] avg 20.67ms, allocated 807 MiB
```

sha c3dd834c008c1a98e71733a321c5e84ba9009269

Reply:
As discussed offline, an updated patch using the latest caching logic and a similar split or enabled/disabled shows 50-90% memory reduction using caching.